increase block-relay-only connections

amitiuttarwar opened the pull request.

mzumsande committed:
net: increase inbound capacity for block-relay-only connections

..and adjust the eviction logic.
The new default max connection number is 200, the default maximum of tx-relaying
inbounds is limited to 60% of all inbound connections.
With 11 outbound connections, that is (200 - 11) * 0.6 = 113.
As a result, the tx-related maximum traffic should not change
drastically.

When we receive an inbound connection and don't have space for another
full-relay peer, we now attempt to evict specifically a full-relay inbound.

Once this commit widely deployed, the added inbound capacity will
allow us to increase the number of outgoing block-relay-only connections.

Co-authored-by: Amiti Uttarwar <amiti@uttarwar.org>

mzumsande committed:
net: add option to evict a tx-relaying inbound peer

If the added bool tx_relaying_peer is set, non-tx-relay
peers will be exempt from disconnection.

Co-authored-by: Amiti Uttarwar <amiti@uttarwar.org>

amitiuttarwar committed:
scripted-diff: Rename connection limit variables

-BEGIN VERIFY SCRIPT-
sed -i 's/nMaxConnections/m_max_automatic_connections/g' src/net.h src/net.cpp
sed -i 's/\.nMaxConnections/\.m_max_automatic_connections/g' src/init.cpp src/test/denialofservice_tests.cpp
sed -i 's/nMaxFeeler/m_max_feeler/g' src/net.h
sed -i 's/nMaxAddnode/m_max_addnode/g' src/net.h src/net.cpp
sed -i 's/m_max_outbound\([^_]\)/m_max_automatic_outbound\1/g' src/net.h src/net.cpp
-END VERIFY SCRIPT-

Co-authored-by: Martin Zumsande <mzumsande@gmail.com>

amitiuttarwar committed:
doc: improve documentation around connection limit maximums

Co-authored-by: Martin Zumsande <mzumsande@gmail.com>

amitiuttarwar committed:
net, refactor: move calculations for connection type limits into connman

Currently the logic is fragmented between init and connman. Encapsulating this
logic within connman allows for less mental overhead and easier reuse in tests.

Co-authored-by: Martin Zumsande <mzumsande@gmail.com>

amitiuttarwar committed:
net: add m_max_inbound to connman

Extract the logic for calculating & maintaining inbound connection limits to be
a member within connman for consistency with other maximum connection limits.

Note that we now limit m_max_inbound to 0 and don't call
AttemptToEvictConnection() when we don't have any inbounds.
Previously, nMaxInbound could become negative if the user ran with a low
-maxconnections, which didn't break any logic but didn't make sense.

Co-authored-by: Martin Zumsande <mzumsande@gmail.com>

amitiuttarwar committed:
test: add functional test for inbound maxconnection limits

Co-authored-by: Martin Zumsande <mzumsande@gmail.com>

amitiuttarwar commented:
- addressed martin's two review comments: updated the sed command in the scripted diff commit message & added the `std::max(0,...)` guard for the inbound max

amitiuttarwar commented:
two updates to the scripted diff 
* added the `m_max_outbound` -> `m_max_automatic_outbound` rename to scripted diff using `/1` trick 
* changed the script slightly to be less confusing around `nMaxConnections` -> `m_max_automatic_connections`

mzumsande commented on src/net.h:
```diff
@@ -75,7 +75,9 @@ static const int MAX_FEELER_CONNECTIONS = 1;
-static const unsigned int DEFAULT_MAX_PEER_CONNECTIONS = 125;
+static const unsigned int DEFAULT_MAX_PEER_CONNECTIONS{200};
+/** Percentage of inbound connection slots that tx-relaying peers can use */
+static const float FULL_RELAY_INBOUND_PCT{0.6};
```
No idea why windows CI runs on your repo, but it's nice. It complains about the float - will change it to double later.

mzumsande replied:
actually changed it to int because it's a percentage.

amitiuttarwar committed:
init: make inbound tx relay percentage configurable

Permit users to change the amount of inbounds that are permitted to relay
transactions. This is particularly relevant to ensure that superusers that are
not concerned with resource usage are not artificially restricted from offering
many transaction relay slots to the network.

Co-authored-by: Martin Zumsande <mzumsande@gmail.com>

mzumsande committed:
p2p: trigger possible eviction if we support bloom filters and change a peer to tx relay

Co-authored-by: Amiti Uttarwar <amiti@uttarwar.org>

mzumsande commented on src/net.h:
```diff
@@ -824,7 +825,7 @@ class CConnman
         m_max_automatic_outbound = m_max_outbound_full_relay + m_max_outbound_block_relay + m_max_feeler;
         m_max_inbound = std::max(0, m_max_automatic_connections - m_max_automatic_outbound);
-        m_max_inbound_full_relay = std::max(0, static_cast<int>(FULL_RELAY_INBOUND_PCT * m_max_inbound));
+        m_max_inbound_full_relay = std::min(1, std::max(0, static_cast<int>(connOptions.m_full_relay_inbound_ratio * m_max_inbound)));
```
the added `min(1, ...)` seems incorrect (and breaks the test). It means we can't ever have more than one FR inbound?

mzumsande commented:
So I made a few changes:
- fixed `p2p_eviction.py` test (in a separate commit for easier review, will squash later)
- changed `FULL_RELAY_INBOUND_PCT` to an `int` (because it's a percentage)
- in commit WIP: update so ratio is init param: clamped `m_full_relay_inbound_ratio` it to a range between 0 and 100 in init (and removed the std::min in `CConnman::init` for that). That way it works and the tests succeed - feel free to improve this, e.g. by error messages if it's out of bounds, or other solutions you had in mind.
- added a commit to handle the bloom filter logic (+test): It wasn't possible to reuse the code from `CreateNodeFromAcceptedSocket` like I planned, because in that situation we haven't connected the new peer yet (and therefore need to check if we are at the limit) and also must check for total peer count, whereas in net_processing, the peer is already connected, and we just have to check after changing tx_relay if the tx-relay count is exceeded. The way I implemented is to first try the normal eviction logic, and if that fails (all peers are protected) we evict the peer that just sent us a filter.

amitiuttarwar commented on src/net.cpp:
```diff
@@ -1698,6 +1698,23 @@ int CConnman::GetExtraBlockRelayCount() const
 
+bool CConnman::EvictTxPeerIfFull()
+{
+    int tx_inbound_peers{0};
```
can this function return early if `fListen` is false?

mzumsande replied:
yes - I think we can just not call it in `net_processing` if the bloom peer isn't an inbound connection - will push an update for that.

amitiuttarwar commented on src/net.h:
```diff
@@ -927,6 +927,13 @@ class CConnman
     int GetExtraBlockRelayCount() const;
 
+    /**
+     * If we are full, attempt to evict a tx-relaying peer
```
maybe: If we are at capacity for inbound tx-relay peers, attempt to evict one.

mzumsande replied:
yep, that's better!

amitiuttarwar commented on src/net_processing.cpp:
```diff
@@ -4770,6 +4770,11 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
             pfrom.m_relays_txs = true;
+            if (!m_connman.EvictTxPeerIfFull()) {
+                // We don't have room for another tx-relay peer, disconnect
+                LogPrint(BCLog::NET, "filterload received, but no capacity for tx-relay and no other peer to evict. disconnecing peer=%d\n", pfrom.GetId());
```
typo: disconnecting

amitiuttarwar commented:
pushed changes to clean up the init param commit. small things like changed the variables to match the percent/int pattern, rewrote commit message, etc.

mzumsande commented on src/init.cpp:
```diff
@@ -1727,6 +1728,7 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
     CConnman::Options connOptions;
     connOptions.nLocalServices = nLocalServices;
     connOptions.m_max_automatic_connections = nMaxConnections;
+    connOptions.m_full_relay_inbound_percent = std::clamp<int>(args.GetIntArg("-inboundrelaypercent", DEFAULT_FULL_RELAY_INBOUND_PCT), 0, 100) / 100.0;
```
now that it's renamed from ratio to percent. we should keep it an int, and do the division with 100 just in connman init.

mzumsande replied:
moved the division.

mzumsande commented:
[56d9caf ](https://github.com/amitiuttarwar/bitcoin/commit/56d9cafe8b36125b8459e738f46d6d76fc87759e)to [460c974](https://github.com/amitiuttarwar/bitcoin/commit/460c974a803f01eedb753e10ad76bfc05ef7eb59) : addressed feedback, also pushed for CI. Will push once more at least.

mzumsande commented on src/net.h:
```diff
@@ -1197,19 +1197,25 @@ class CConnman
     std::unique_ptr<CSemaphore> semAddnode;
-    int m_max_automatic_connections;
-
-    // How many full-relay (tx, block, addr) outbound peers we want
```
did you drop the doc for  `m_max_outbound_full_relay` and `m_max_outbound_block_relay` on purpose?

mzumsande commented:
some more small changes:
commit "net: add m_max_inbound to connman"
expanded commit message, dropped "refactor" from it because of the added `max()`

commit "doc: Cleanup documentation around connection limit maximums"
added to m_max_automatic_connections the info that it includes inbounds

commit "net: increase inbound capacity for block-relay-only connections"
updated commit message

commit "test: add functional test for inbound maxconnection limits"
fixed possible timeout, changed doc a bit
